### DISCUSSION TITLE
Real PITA that I am willing to tackle.

### DISCUSSION THREAD

**Discussion by KC5VDJ** [kej715/DtCyber#145] [General]
I want to drop stdio for anything but host-side messaging....

AAAAARRRRRRGGGGHHHHH!!! How long to itemize three tapes? disk as well... in 2024, the sweet spot is closer to 4 megs to 16 mega per block transferred. all but the tightest pi zero original, and EVEN THEY TOO (maybe), should be able to do at least a single buffer for a single host block at max efficiency block size, say in types...defaulting to, say, 8 megs... the biggest io bottleneck in dtcyber currently is the very portable, but usually pawned off to the os directly, stdio package. the solution is obviously lower level file io and diy buffering. It will be as SSD friendly as possible (or not, depending on selections)...most people have a battery based or backed platform..reduced risk of catastrophic failure.

How big a block should be in types.

There is a time and place for everything, but stdio to simulate physical primary devices of a computer system...  Not exactly it's purpose.  This why K&R added low-level (open/read/write/select, etc).

**Comment by KC5VDJ** [comment]
I'll have to pre-define the framework of course, so that any other efforts can adapt pre-pub.

I anticipate significant increases in io speed.  No, don't ask for a prediction, because that is os-dependent, even compiler dependent.  Still, I would anticipate SIGNIFICANT INCREASES for whomever plays with block size.

Accounting for non-battery-backed systems, would be a less SSD friendly option, but still better than stdio and tiny blocks.

**Comment by kej715** [comment]
Would this be an exercise in "premature optimization"? Are you really sure
that stdio is the bottleneck and that eliminating its usage would make a
significant difference? Remember that most I/O is performed a byte at a
time across emulated channels to/from PP's, and that one byte is
transferred per PP cycle. In addition, tape I/O tends to be more complex
than, e.g., disk I/O. It seems more likely that the I/O bottleneck is due
to the mainframe architecture than to the emulator's usage of stdio.

Compare the performance of the emulated 885-42 disk subsystem to the
performance of the ordinary 885 subsystem. The 885-42 performs
significantly better than the ordinary 885. This is because the 885-42 is a
DEMA (Direct Extended Memory Access) subsystem. The 885-42 transfers sector
data directly to/from ESM instead of transferring it a byte at a time over
a channel, and it uses the CPU to perform block transfers between ESM and
jobs requesting the data. The 895 disk subsystem was similar in operation
to the 885-42 except that it was DMA instead of DEMA. Unfortunately, CDC
did not provide a tape subsystem that supported DEMA or DMA.

From an historical perspective, consider that itemizing three full tapes
would have taken quite a long time on the real hardware. It's very likely
that DtCyber already does it much faster than the original hardware could.

**Comment by KC5VDJ** [comment]
Good points...  Still, I'm at least going to experiment.  If the improvements are significant enough....we'll see, I guess.  Most stdio implementations kinda suck though for such things, suckiness variable by compiler/os...

If the hosted/simulated side is part of it, and it plays it's part in this, at least the host-side inefficiency would be somewhat mitigated.

**Comment by KC5VDJ** [comment]
Hmmm....  Full build as benchmark.

Reliably IO-bound.  Long run...long enough.

**Comment by nickglazzard1959** [comment]
(Don't know why this is in bold - not intentional!) I'm not sure how useful it would be, and maybe you have already done this, but it is possible to profile DtCyber fairly easily, at least on Linux. I haven't done this sort of thing for a very long time (and that was mostly on IRIX), but if you rebuild DtCyber with -g -pg and remember not to strip symbols with -s when linking (and use -pg then too), you can get function usage information from gprof. E.g.
`sudo ../dtcyber (in NOS2.8.7)`
`sudo gprof -b ../dtcyber gmon.out > deadshut.txt `
gives this sort of information:
---
```
Flat profile:

Each sample counts as 0.01 seconds.
  %   cumulative   self              self     total           
 time   seconds   seconds    calls  ms/call  ms/call  name    
 21.15     62.19    62.19 241662556     0.00     0.00  ppStep
 17.58    113.90    51.71 1158027553     0.00     0.00  cpuStep
 11.33    147.23    33.33 2007119421     0.00     0.00  cpOp47
  8.81    173.15    25.92 241113812     0.00     0.00  channelStep
  4.33    185.89    12.74 1110788812     0.00     0.00  cpuFetchOpWord
  3.36    195.78     9.90 2328564603     0.00     0.00  cpuAddRa
  3.07    204.81     9.03 1624099176     0.00     0.00  ppOpNJN
  2.59    212.43     7.62  8348103     0.00     0.00  ppOpADN
  2.54    219.90     7.47 1098205989     0.00     0.00  cpuVoidIwStack
  2.47    227.17     7.27 1206362140     0.00     0.00  cpuCheckOpAddress
  2.25    233.78     6.61 1021867707     0.00     0.00  cpOp04
  2.16    240.14     6.37 60018989     0.00     0.00  ppOpZJN
  1.53    244.64     4.50 1835661019     0.00     0.00  ppSubtract18
  1.35    248.61     3.97 398488656     0.00     0.00  idleThrottle
  1.34    252.54     3.93 1797345202     0.00     0.00  ppOpSBN
  0.63    254.38     1.85 200161644     0.00     0.00  ppOpIAM
  0.59    256.12     1.74 46929764     0.00     0.00  ppOpCRM
  0.56    257.77     1.65 234512245     0.00     0.00  ppOpPJN
  0.52    259.31     1.54 171066883     0.00     0.00  ppOpLDD
  0.51    260.82     1.51 76447057     0.00     0.00  ppOpCRD
  0.51    262.32     1.50 123354829     0.00     0.00  cpuPpReadMem
  0.49    263.76     1.44 159070300     0.00     0.00  dd8xxIo
  0.47    265.13     1.37 429301356     0.00     0.00  idleDetectorNOS
  ...
  ---
  ```
  there is a lot more. This is for deadstarting, logging in once then shutting down (on a rather old version: 5.7.38).
  I think you can get more detailed info by source code line using gcov (this seemed easier on IRIX, TBH!).
  Anyway, it might be worth doing some profiling to see where time is being spent and hence where effort improving things would best be spent too. 
  I have a feeling there won't be any single standout thing that can easily be improved, otherwise Tom and Kevin would probably already have done it.

**Comment by wschaub** [comment]
Aside from any performance gains you might achieve  Wouldn't this complicate support for windows? I can't imagine any low level IO APIs you could use on a POSIX like system would also be on windows.

**Comment by KC5VDJ** [comment]
open(2)/read(2)/write(2)/select(2), etc., have been around since K&R 1st Ed, and to my recollection are even in most (all?) Tiny-C implementations.  Classical straight C.  MickeySoft has had them forever.

**Comment by kej715** [comment]
Correct. If stdio is abandoned, it will become necessary to start dealing
directly with Windows file handles, and the lower-level file I/O API's on
Windows are significantly different than POSIX open/close/read/write/lseek,
etc. The code will need to have many more occurrences of "#if
defined(_WIN32)", or a stdio-like file I/O abstraction module will need to
be implemented.

The quick profiling report produced by Nick seems to confirm that stdio is
not actually a bottleneck. In the snippet he provided, stdio would be
called only from within the dd8xxIo module, and that accounts for less than
0.5% of the total time consumed, which is also less than the time consumed
by the PP IAM instruction emulation (i.e., the instruction that is pulling
one byte at a time over the emulated channel).

**Reply by KC5VDJ** [reply]
I was thinking of the abstraction for minimum change, keeping it close, but allowing for compile-time buffer sizing.

**Reply by KC5VDJ** [reply]
Oh, and yes.  stdio is stdio, standard, portable...guaranteed, so to say.  The ifdefs will largely be in the device dependent implementation files as currently done for other purposes.  Some weird sh.. has been happening to c in some of these calls as well, blame black hat...

The idea being as little change as possible to the original sources.  I'm still not convinced that small block transfer can be handled efficiently outside of the program doing it, and even then, using large buffers.  Operating systems don't have the info to really do any optimization on this.  AI techniques have potential, but would have to be simple, because of its purpose, a small net maybe in hardware..  It ain't there yet.  I don't feel like waiting.

I see doing it this way, as an experiment, as safe, not as big a pita as I was thinking (even for everyone else), and if it can do 10-??% improvement, worthy of the LOE, and inclusion.  I put the cutoff at 10% performance increase, less if part of a larger optimization with serious improvement.  On that, as Nick was saying, although the CPU instructions are well written, there could be cause to look there as well.  (Face it, TK5 sucks, the only thing it has going for it with it's Apollo toolset, with the exception of the inclusion of gcc (kudos), is ispf.  Let's make it suck worse!  The hobbyist thing).

Also, there is a motivation to also make it Intel -ipo compatible (compile time multi-source optimization and self-vectorizing), and in that process get rid of some of those speculative execution holes maybe (LOE and duplication of effort. The one justifies the other, because they are interrelated, with similar LOE. Kill two birds).

**Comment by KC5VDJ** [comment]
Hmmm...  Been a while since I did the laptop.  Interestingly, the sweet spot for the 860 EVO 500GB is one meg.  Also interestingly, RAPID is supposed to only be on the boot drive, according to Magician, but, the results from the second drive really look like it is caching there.

![cQltg5I8gn_103400](https://github.com/kej715/DtCyber/assets/136194499/3e7fb66d-4a44-467c-a3d8-d665bea3b863)

**Comment by nickglazzard1959** [comment]
One curious thing from that quick profile. cpOp47 is using 11% of the run time on this deadstart/shutdown run. That is the population count instruction! Can this be right? :-) If it is, it might be worth fiddling with. It is a clever algorithm already, but a book called "Hacker's Delight" by Henry S. Warren may have an even faster one ... He has a whole chapter on "bit counting" methods. 
Hmm ... actually the Wikipedia article on "popcount" contains the same clever algorithms as that book, so they are easy to find ...

**Reply by KC5VDJ** [reply]
An overall spring cleaning and tuneup would be a good thing.

**Comment by kej715** [comment]
The CPUMTR idle program spins in a tight loop executing population count
instructions:

                                       **        IDL - IDLE PROGRAM.
                                    CPUMTR   14527

                                   CPUMTR   14528

                                   CPUMTR   14529

                                   CPUMTR   14530
         7673  00000000000000000000     IDL      CON    0           (RA)
FOR IDLE PROGRAM                      CPUMTR   14531
         7674  00000000000000000000              CON    0           (RA+1)
FOR IDLE PROGRAM                    CPUMTR   14532
         7675  47211                             CX2    X1          DELAY
(6 BITS)                             NS2332       6
                    47211                        CX2    X1          DELAY
(6 BITS)                             NS2332       7
                         0400000002              EQ     2           LOOP
                                    NS2332       8

**Reply by nickglazzard1959** [reply]
Hi Kevin, That *might* (just possibly) mean it is worth trying a (perhaps) faster bit count implementation. I might play with that at some point.

**Comment by kej715** [comment]
If a system is idle, so the CPU is spending almost all of its time
executing the pop count instruction, shouldn't we expect that emulation of
this instruction would feature prominently in an emulator execution
profile? During deadstart, the CPU is probably mostly idle because the PP's
are doing almost all of the work. Similarly, after a typical hobbyist
system is up and running, the CPU tends to be mostly idle unless it has
been set up deliberately to run maintenance jobs, or the hobbyist has some
other CPU-intensive job mix running.

After revisiting the code in CPUMTR and being reminded that the idle loop
comprises almost entirely of the pop count instruction, it didn't seem
surprising that the profile would feature this instruction prominently.
And, the current implementation of pop count in DtCyber looks amazingly
efficient. Compare it to implementation of the XJ instruction or the
UEM/ESM block move instructions, for example. More power to you if you
would find improving pop count personally challenging and rewarding (it's
certainly one of the machine's most historically unique features), but I'm
not sure of the holistic value (beyond personal satisfaction) of improving
the machine's performance mostly when it's idle. ;-)

Note that pop count continued to evolve and live on in other Cray machines
such as the CDC 8600, Cray 1, Cray X-MP, etc.

**Reply by nickglazzard1959** [reply]
:-) ... yes, improving the machine's performance while it is idle may not be the most useful thing to do. Good point! Looking at the "Hacker's Delight" book, it looks like it could be even faster (although it is certainly pretty clever as it is), but ... yeah ... :-) 
Pop count is certainly a famous Cray system feature , reputedly of great interest to certain 3 (and 4) letter agencies ... :-)

**Reply by KC5VDJ** [reply]
Well, although it could be used for parity, other hardware handled that.

Consider the initial and biggest customers.  Obviously NSA.  DOD/CIA/NRO/AEC for various.  All sorts of uses.

You'll also find Flash and Flash Override in the TCP/IP stack in NAM.

**Comment by KC5VDJ** [comment]
Just looking at grep -n FILE *.h *.c

I'll leave debug logs on stdio.  If there is a time and place.

**Comment by KC5VDJ** [comment]
Oh, and maybe I'm wrong.  Fuggit.  Worth a try.  I may be right and wrong at the same time, hardware/os specific.  Still worth a try.  Then again, maybe I am right and minimum standard means minimum, and most operating systems kinda suck in the use-case of stdio, and this gets notable improvement.  Still worth a try.

**Comment by KC5VDJ** [comment]
Targets of experiment:

dd8xx.c
dd885-42.c
mt679.c

And associated includes.

Test environment:

Eleven year old Haswell laptop, 16 gigs not crammed, win10 current updates.  Linux and BSD available (need to get that up to date), but testing will be on winblowz.  Intel compiler, max compatible optimization, incl. Hardware specific, AVX2 (universal now on Intel/AMD), self-vectorizing (max compatible), and single-file optimization (DtCyber is not currently -ipo compatible).

Threshold for "success":

10% improvement.  Tell me that isn't enough after an n-hour run.

Caveat:

If part of a much larger spring cleaning and tune-up, a smaller percentage MIGHT be acceptable, depending on contribution to aggregate improvement.

Acceptable?

**Reply by KC5VDJ** [reply]
It also has to have as minimal/tunable impact on a half gig system as possible.  I see no need to drop any support, short of going the FreeBSD route and fine-tuning for specific architectures.  The latter being the Oxford definition of PITA as things stand currently.

**Comment by kej715** [comment]
I just completed some profiling of DtCyber on MacOS using the Xcode
profiler. Based upon these results, combined with Nick's previous gprof
results on Linux, it seems that eliminating stdio would not yield any
measurable improvement in overall performance. I ran each of two
experiments a number of times and saw the same results for each run. The
experiments were:

Experiment 1 - Run the following job and profile DtCyber while the job is
running. The job catalogs the NOS 2.8.7 deadstart tape three times. This
should expose tape I/O performance.

TEST1.
USER,INSTALL,INSTALL.
ASSIGN,50,TAPE,LB=KU,F=I.
CATALOG,TAPE,N,R.
CATALOG,TAPE,N,R.
CATALOG,TAPE,N,R.


This produced results similar to the results reported by Nick.
Specifically, ppStep consumed about 40% of the time. Within this, the IAM
instruction consumed about 4% of the total time. Within that, the channelIn
function consumed 0.1% of the time, and mt679 (the module that actually
calls stdio) did not consume any measurable amount.

Experiment 2 - Run the following job and profile DtCyber while the job is
running. The job copies the NOS 2.8.7 deadstart tape to a local file twice
with verification, so the tape is read and copied to a file, then both are
rewound, read and compared. This is about as I/O intense as you can get and
should expose a combination of tape and disk I/O performance.

TEST1.
USER,INSTALL,INSTALL.
ASSIGN,50,TAPE,LB=KU,F=I.
COPY,TAPE,DS,V.
COPY,TAPE,DS,V.

In this case IAM consumed about 4.5%, and OAM showed up and consumed 1.3%.
mt679 consumed 0.8% and dd8xx consumed a total of about 0.6% (sum of input
and output). In this case, the profiling actually started to report that
stdio functions such as fread and fwrite were being noticed. However, they
were reported as consuming 0.0%, so they were consuming less than 0.1%.

Thus, even if replacing stdio in mt679 and dd8xx would improve their
performance by as much as 10% (this seems unlikely given profiling
results), that improvement would not have a measurable impact on I/O
performance as a whole. The end result would be a custom-built abstraction
module that we would need to continue maintaining. It would have equivalent
functionality to stdio with almost no measurable performance benefit.

Have you completed your implementation of the drum I/O module? Have you
thought about doing something like implementing support for the 895 disk
subsystem? That would be a very nice contribution to DtCyber. The 895 was
one of the last high performance, buffered I/O disk subsystems supported by
NOS on Cyber 800 systems. Presumably, NOS 2.8.7 would support it on the
865/875.

**Reply by KC5VDJ** [reply]
Interesting.  One would think such an abstraction module would...but...there it is.  Good points.

And given some drive upgrades, I might be able to find some space for it.

My biggest problem right now is time.

Oh, and the angle translation will be at init and dynamic.  No sense in huge static table includes, even though they could be, they get built at init.  Made the 13/14/15 bit rotates a bit more library-like, different variations, as separate functions (fastest clobbers source high bits, slowest protects everything, could come in handy).  The same can be easily adapted to bit fields (increased cycles cost, as relocation to low bits would be necessary, otherwise, the same).  An address translation table being the obvious solution, well, indirection table (makes interleave compatibility inherent, but requiring specification in cyber.ini; or I can just ignore interleave in DtCyber, as long it is there logically; either way, the indirection table is required.  Implementing actual interleave would be trivial, but possibly/probably pointless, and is already there on real steel.  Unless someone finds a drum and might want an image to load, again, would be trivial.  I would imagine that most actual use cases at any point in time (not counting actual drum machines) have no need for such images.

Unless someone can think of a historical use-case that would have relevance, at the same time requiring a DtCyber generated image, interleave will be implemented as logical-only.  The one is a function of the other, if it is so trivial, it may go in anyhow.  Ability to specify interleave in the dt console could be handy, but increases either dimensions at dt init or later time to recompute just one.  Kinda stuck at the moment on that direction.  I try to keep pi zero in mind, gotta have room for other stuff.  Not '82.  Half a gig is tight.

**Reply by KC5VDJ** [reply]
Any known cases of multiple interleaves on one drum?  Two controllers could access one physical drum.  I would think not in most use cases, but....realtime/nrt?  Given the customer list, I see scenarios, just rare.  Whichever, it still has to be compatible with real steel.

**Comment by nickglazzard1959** [comment]
Thanks for doing the profiling! The results seem to be very much what we might expect. I don't know if you would agree entirely, but at first glance, nothing "stands out like a sore thumb" as a single thing that would significantly increase performance if it could be improved. (I got mistakenly excited about pop count because it looked like it might be that). 
I would certainly be interested personally in an implementation that was arbitrarily fast, completely ignoring history ... :-) ... but it looks hard to do much better than we are already doing without using other techniques which are probably only feasible with a team of engineers working on it full time. I'm thinking of simulators/emulators using code translation. This is completely irrelevant, but Bob Supnik of SIMH fame was leading a team at Unisys about 12 years ago that created a Burroughs B65xx (?) simulator for MCP running on Xeons. It reached 550 MIPS of the simulated architecture in 2012. This proved that high performance simulation was possible even when the target and host architectures were very different indeed (48 bit word with tags, stack based, etc. to x86-64). AIUI, this is what the Unisys "mainframes" are: advanced simulators running on Xeons. There is no actual Unisys hardware.
Such pipe dreams aside, I'm very happy with the current performance of DtCyber running NOS. I'd guess it is as fast or faster than the 170-865 hardware was (probably much faster at some things), at least if run on reasonably fast hosts (I found RPi4 just about OK with NOS 2.8.7, Odroid XU4 was significantly better, an old Mac Mini or NUC ideal). Running the Cray X-MP simulation at real hardware speeds would probably require the code translation approach, although I haven't really made measurements TBH.

**Reply by KC5VDJ** [reply]
Such a project would be a split into a more FreeBSD philosophy (arch-specific focus).

Such would allow for lower level solutions than just standard c.  HLL as well, as at least Intel has an actual HPF, and after the '24 language standards are finalized after the OpenMP conference this summer, integrated OpenMP vectorization support in most languages of note.

Oh, and such would be the Oxford definition of PITA.

There are a lot of small things that could be done to speed things up, but only worth the effort for the cumulative improvement if done at once.  The CPU code, as mentioned is quite well written, and other than new algorithms to accomplish the same, is quite efficient.  There is stuff that could be optimized though.  Arch-specific could produce results, but...could desynch things badly across architectures without a committed team/individual per architecture coordinating with and as a committee/core group across architectures (buy a hat rack).

I honestly would love to, but as I said to Kevin, I'm trying to find to time to get more than fragments of my current stuff worked on.  A better situation could change that.  In the meantime, DtCyber kicks butt on TK5, although TK5 has ispf and gcc to its favor, but NOS 2.8.7, even 1.x has a more modern toolset, excepting the gcc bit.

**Reply by KC5VDJ** [reply]
Oh, and what Kevin said.  If NOS itself is the bottleneck.  That'll produce limited results.   At 1,179,986 lines of mainly compass, just for the opl871 I am running (unless I missed something and it's more), automated tools for a lot of the analysis is probably the best way to go.  I have ideas on some of it, but a true big picture with easy drill-down would require a PC based assembler/disassembler with full compass compatibility, dumping the tables for visualization.

What is it?  Over four million lines?  How do you do an exhaustive analysis for such in a small team in random spare time?  Automate as much as possible 

My other question kinda would require it.  Which bits are truly available for future NOS, and not marked for "installation reserved". The idea being to document the undocumented and usable.  Unless someone has the os team's documentation at that level.  Anecdotal evidence says it is not available (even if to some.  See customer - list ).  BE was analyzed to that level by a man named Tracy.  Welcome to '24.  Better tools.

**Reply by KC5VDJ** [reply]
I kinda like the idea of os construction kit.  And the anecdotal evidence shows THEY dropped a hit and watched three episodes of Star Trek together (that one report by the CDC/NCR team).

An analogy would be a trusted code base or standard cells.  NOS lends itself to that actually, but only if such a breakdown of the code is done first.  A lot of it.  Not overwhelming, but time-consuming.

That CDC/NCR team could claim a form of prior art on H2G2 being self-creating (Douglas Adams).

"BRAIN? BRAIN! WHAT IS BRAIN!" Season 3.  In the top three worst episodes.  "Help us Landreau!" (Season 1)....

**Comment by kej715** [comment]
To clarify ... NOS is _not_ the bottleneck. Our profiling is illustrating
how the machine architecture operates and why the performance of stdio
doesn't greatly influence emulation of it, at least with regard to most
peripheral equipment. Consider how I/O works for most types of disk
subsystems on a CDC 6000 series machine, regardless of operating system.
For example, after positioning to a specified track and sector, a PP driver
issues an IAM instruction to read the contents of the sector. For most disk
types, the IAM is set up to read one PRU which is 320 12-bit bytes. These
bytes are read one by one across the channel into PP memory. One byte is
read each PP cycle. After reading the 320 bytes in 320 cycles, the PP
driver will then typically transfer the PRU to a job control point in
central memory, e.g., using a CWM instruction. CWM will transfer 5 bytes (1
60-bit word) per PP cycle. Thus, reading a sector from a file to a job's
field length requires 384 PP cycles (i.e., 384 trips around the emulated PP
barrel). That's how the original hardware worked too.

The one peripheral that doesn't operate as described above is the 885-42
disk subsystem. The controller for this disk subsystem connected directly
to an ESM port. The PP driver uses a Cyber channel to send functions to the
controller, but it doesn't use the channel for transferring data. Instead,
when it wants to read/write sectors, it sends a function that causes the
controller to read/write sectors directly to ESM via the directly connected
port. Code in CPUMTR then uses block move instructions to transfer the data
buffered in ESM to/from control points in central memory.

The 885-42 had another interesting feature that made it a very
high-performance device, and that compelled its DEMA (Direct Extended
Memory Access) design. It read/wrote four heads in parallel. Thus, instead
of transfering a single sector for each read/write request, as most other
disk devices did, it read/wrote four in parallel, quadrupling the number of
12-bit bytes that were transferred per unit of time, and outpacing what a
PP could handle over a normal channel. And that's why the controller was
designed to connect directly to ESM and transfer data directly to/from
there.

**Reply by KC5VDJ** [reply]
That last bit is similar to the way the 865 drum did.  Sector-based.  Spreading the bits across three tracks, although the layout suggests a rather simple(*) hardware reassembly of the words (parallel in serial out shifter) on the drum.

\* As "simple" as a 12 bit parallel in serial out shifter can be using discrete germanium.  (TO-5 like IBM?)

**Reply by KC5VDJ** [reply]
In the early 90's I specified some smaller parallel drives for a video system.  They were the only thing available at the time at any reasonable price that could do realtime 24/30 frame at depth/res...

**Reply by KC5VDJ** [reply]
And yes, this laptop peaks at 22-24 GFLOPS double, with 1700-2400 MFLOPS on the low end doing a lotta divides (flops2.c, nosc.mil, Intel compiler, avx2).  170-190 GFLOPS in aida64's gpgpu (I call bullshit for anything realistic).  So, yeah, I get that.  This not a high performance CPU anymore, but could be close to the average form of dedicated or semi-dedicated system running this.  Most hobbyists I would imagine will be running it on old machines they don't want to trash yet.

It is still a simulation.
